Adds TLS requirements for issuer keyset (#176)
smart-on-fhir/health-cards: docs/index.md

The public key listed in the first certificate in the `"x5c"` array SHALL match the public key specified by the `"crv"`, `"x"`, and `"y"` parameters of the same JWK entry.
If the issuer has more than one certificate for the same public key (e.g. participation in more than one trust community), then a separate JWK entry is used for each certificate with all JWK parameter values identical except `"x5c"`.

> Issuers SHALL publish their public keys as JSON Web Key Sets (see [RFC7517](https://tools.ietf.org/html/rfc7517#section-5)), available at `<<iss value from JWS>>` + `/.well-known/jwks.json`, with [Cross-Origin Resource Sharing (CORS)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin) enabled, using TLS version 1.2 following the IETF [BCP 195](https://www.rfc-editor.org/info/bcp195) recommendations or TLS version 1.3 (with any configuration).

Reviewer: Thanks -- "with any configuration" heads off the question I had when I first read this :-)

Issuers SHALL publish their public keys as JSON Web Key Sets (see [RFC7517](https://tools.ietf.org/html/rfc7517#section-5)), available at `<<iss value from JWS>>` + `/.well-known/jwks.json`, with [Cross-Origin Resource Sharing (CORS)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin) enabled.

> The URL at `<<iss value from JWS>>` SHALL use the `https` scheme, and SHALL use the following TLS version 1.2 or 1.3 configurations:

Reviewer 2: I'm uncertain of the intent of this sentence. If a webserver support TLS 1.0, 1.1, 1.2, and 1.3, is that webserver in compliance with this requirement? Could you please clarify?

Generally, are you trying to define a floor or a ceiling? They proposed phrase: "support at least one" instead of "use", below, implies a floor, such that the connection could be insecure if the client accesses the public key doesn't also support these best practices / requirements. 

This PR really begs the question -- why are we putting normative requirements on the server instead of the client?

Reviewer: Re: guidance for servers vs clients: if a server doesn't meet these requirements, *nothing* a client does can achieve an appropriate level of protection. So these guidelines are intended to ensure that clients *can* obtain keys in a secure fashion. Clients don't need to retrieve keys directly from an issuer (e.g., they could rely on an intermediary), so I'm not sure it makes sense to impose additional constraints on that side of the interaction.

Re: "ceiling" vs "floor", these terms can be a bit confusing; I *think* the intent of this langauge is to ensure at least one of TLS 1.2 or 1.3 is supported. (Possibly we also want to prohibit servers from supporting TLS 1.0, but that's another question.)

As such, I'd suggest re-wording as:

```suggestion
The URL at `<<iss value from JWS>>` SHALL use the `https` scheme, and SHALL support at least one of the following TLS configurations:
```

Author: Superseded by ref to BCP 195 (in line with the "at least" suggestion)

> * For TLS 1.2, use a ciphersuite conforming to the following algorithm choices and _minimum_ cryptographic strength:

Reviewer: ```suggestion
* For TLS 1.2, support at least one ciphersuite conforming to the following algorithm choices and _minimum_ cryptographic strength:
```

Author: Meaning we're ok if the endpoint supports less secure options, allowing non-conforming clients to still get the keys?

Reviewer: I think that's consistent with our internal discussion, yes -- here, I mainly intended it as a disambiguation of the way I already read "use a ciphersuite conforming".

Let me know if you think this is the wrong call.

Author: Given this is a new spec leading to (presumably) new implementations, enforcing a secure min bar makes sense to me. Should be ok for servers, even if keys are hosted by legacy systems. Clients (wallets and newly created apps) should also support this min bar. Bottom line: which legacy components that can't meet these recommendations are we trying to support?

Reviewer: I think the potential issue is with existing web servers that support TLS 1.1 *in addition to* supporting TLS 1.2 with ciphersuites we recommend. The question is whether that's considered OK or not. Agree it's worth discussing on a call (I expect @isaacvetter may want to join).

Author: Superseded by ref to BCP 195 (TLS 1.1 is now deprecated)

  * Key exchange: ECDHE using a 256-bit curve
  * Authentication: ECDSA with a 256-bit curve, or RSA with a 2048-bit key
  * Session cipher: AES with a 128-bit key

>   * Cipher mode: CBC or CGM

Reviewer 3: ```suggestion
  * Cipher mode: CBC or GCM
```

Author: Thanks, fixed.

  * Hash algorithm: SHA256
  * The following recommended ciphersuites adhere to these requirements:
    * TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
    * TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
    * TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
    * TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
    * TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256
    * TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384
    * TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256
    * TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384

> * For TLS 1.3, use one of these ciphersuites:

Reviewer: ```suggestion
* For TLS 1.3, support at least one of these ciphersuites:
```

Author: Now allows for any TLS 1.3 config.

    * TLS_AES_256_GCM_SHA384

>     * TLS_AES_128_GCM_SHA256

Reviewer 2: ```suggestion
    * TLS_AES_128_GCM_SHA256
    * TLS_CHACHA20_POLY1305_SHA256
```
Per [rfc8446](https://datatracker.ietf.org/doc/html/rfc8446#section-9.1), why shouldn't `TLS_CHACHA20_POLY1305_SHA256` be in this list?

Reviewer 4: chacha is not a fips 140-2 approved algorithm so it would likely be a problem for federal entities to require support

Author: I kept the proposal to the NIST standardized crypto algs, but we can add the CHACHA20_POLY1305.

Author: Superseded by ref to BCP 195.

> * X.509 certificate SHALL use a 256-bit ECDSA or 2048-bit RSA key, must be valid, and trusted by major web browsers.

Reviewer: ```suggestion
* X.509 certificate SHALL use at least a 256-bit ECDSA or at least a 2048-bit RSA key; SHALL be valid; and SHALL be trusted by major web browsers.
```

These are meant as minimums, right? (There's no outer bullet stating such at this point in the bullet list, so we might want to say so explicitly inline.)

Reviewer 4: reword? The way the last phrase of the proposed change is worded suggests that this spec intends to dictate what 'major' web browsers shall or shall not do; also there is likely disagreement on what defines a 'major' browser, so that is not really enforceable/testable.

Author: Do you have a proposed wording? I was avoiding being explicit for X.509 requirements on purpose here, because this would require _a lot_ of text (what chain properties are valid? Rooting to what? Expiration? Revocation?) Browsers typically have a good handle of what a "valid" cert is; if you use a 'major' one to visit the completed issuer URL and get a green check mark, you are compliant; so could we phrase that in a vendor-neutral way?

Author: And RE: ECDSA vs EC, an auth X.509 cert can only have ECDSA alg, no? RSA is also a sig alg (confusing in this case, because it's name the same if used for encryption or signing).

Author: Superseded by ref to BCP 195
